ksql-engine/src/main/java/io/confluent/ksql/security/KsqlAccessValidatorCache.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.security;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Ticker;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.confluent.ksql.exception.KsqlTopicAuthorizationException;
+import io.confluent.ksql.util.KsqlConfig;
+
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.kafka.common.acl.AclOperation;
+
+/**
+ * An implementation of {@link KsqlAccessValidator} that provides authorization checks
+ * from a Cache.
+ */
+@ThreadSafe
+public class KsqlAccessValidatorCache implements KsqlAccessValidator {
+  private static final boolean ALLOWED = true;
+  private static final boolean DENIED = false;
+
+  static class CacheKey {
+    private static final String UNKNOWN_USER = "";
+
+    private final KsqlSecurityContext securityContext;
+    private final String topicName;
+    private final AclOperation operation;
+
+    CacheKey(KsqlSecurityContext securityContext, String topicName, AclOperation operation) {
+      this.securityContext = securityContext;
+      this.topicName = topicName;
+      this.operation = operation;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == null || !(o instanceof CacheKey)) {
+        return false;
+      }
+
+      CacheKey other = (CacheKey)o;
+      return getuserName(securityContext).equals(getuserName(other.securityContext))
+          && topicName.equals(other.topicName)
+          && operation.code() == other.operation.code();
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(
+          getuserName(securityContext),
+          topicName,
+          operation.code()
+      );
+    }
+
+    private String getuserName(KsqlSecurityContext securityContext) {
```

Comment:
Yes, but I cannot compare the `Optional<Principal>` because the `Principal` may have other variables that are not equal. 

For instance, in RBAC, the `Principal` uses a `JwtPrincipal` class which uses the user name & token to check for equality. The token is different in every session, but the user name is not.

Reply:
Makes sense..